recipe-server/client/control/components/MultiPicker.js

```diff
@@ -0,0 +1,234 @@
+import React, { PropTypes as pt } from 'react';
+
+export default class MultiPicker extends React.Component {
+  static propTypes = {
+    unit: pt.string.isRequired,
+    plural: pt.string.isRequired,
+    options: pt.array,
+    // from redux-form
+    value: pt.oneOfType([pt.array, pt.string]),
+    onChange: pt.func,
+  };
+
+  static getActiveSelectOptions({ options }) {
+    if (!options) {
+      return [];
+    }
+
+    const selected = [];
+    for (let i = 0; i < options.length; i++) {
+      if (options[i].selected) {
+        selected.push(options[i].value);
+      }
+    }
+
+    return selected.filter(x => x);
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      filterText: null,
+    };
+
+    this.handleApplyOption = ::this.handleApplyOption;
+    this.handleRemoveOption = ::this.handleRemoveOption;
+    this.onTextChange = ::this.onTextChange;
+    this.convertValueToObj = ::this.convertValueToObj;
+  }
+
+  onTextChange(event) {
+    const { value } = event.target;
+
+    this.setState({
+      filterText: value,
+    });
+  }
+
+  getDisplayedOptions() {
+    const {
+      options = [],
+      value,
+    } = this.props;
+
+    const selectedOptions = options.filter(option =>
+      value.indexOf(option.value) === -1);
+
+    let displayedOptions = [].concat(selectedOptions)
+      .map(option => option.value)
+      .map(val =>
+        (!val ? null : { ...this.convertValueToObj(val) }))
+      .filter(x => x);
+
+    const {
+      filterText,
+    } = this.state;
+
+    if (filterText) {
+      displayedOptions = displayedOptions.filter(option =>
+        JSON.stringify(option).indexOf(filterText) > -1);
+    }
+
+    return displayedOptions;
+  }
+
+  handleApplyOption(event) {
+    event.persist();
+
+    if (!this.availableRef) {
+      return;
+    }
+
+    const {
+      value,
+      onChange,
+    } = this.props;
+
+    let selectedFilters = MultiPicker.getActiveSelectOptions(this.availableRef);
+
+    if (!selectedFilters || selectedFilters.length === 0) {
+      selectedFilters = this.getDisplayedOptions();
+      selectedFilters = selectedFilters.map(option => option.value);
+    }
+
+    const newOptions = []
+      .concat(value || [])
+      .concat(selectedFilters);
+
+    // clear user input
+    this.availableRef.value = null;
+
+    onChange(newOptions.join(','));
+  }
+
+  handleRemoveOption(event) {
```

Comment:
`handleApplyOption`/`handleRemoveOption` have been refactored down into one function, since they shared near-identical functionality.

As for splitting up the different sides of the MultiPicker, I'm unsure how that would work since we are reading values from `ref`s, and splitting out those components would lose those references. I will investigate.